### PR TITLE
Vk protected content platform

### PR DESCRIPTION
**Pull request by phoenixxxx**
As discussed this is the first phase of content protection.

### CODE REVIEW

File: filament/backend/src/vulkan/platform/VulkanPlatform.cpp [hunk 1]
```diff
@@ -686,10 +706,22 @@ Driver* VulkanPlatform::createDriver(void* sharedContext,
...
     printDeviceInfo(mImpl->mInstance, mImpl->mPhysicalDevice);
 
+    VkPhysicalDeviceProtectedMemoryFeatures queryProtectedMemoryFeatures = {
+    .sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_PROTECTED_MEMORY_FEATURES,
+    };
+    // Chain to the pNext linked list
+    queryProtectedMemoryFeatures.pNext = context.mPhysicalDeviceFeatures.pNext;
+    context.mPhysicalDeviceFeatures.pNext = &queryProtectedMemoryFeatures;
+    VkPhysicalDeviceProtectedMemoryProperties protectedMemoryProperties = {
+    .sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_PROTECTED_MEMORY_PROPERTIES,
```
Comment (poweifeng) [inline filament/backend/src/vulkan/platform/VulkanPlatform.cpp:716]: add 4 spaces here

File: filament/backend/src/vulkan/platform/VulkanPlatform.cpp [hunk 2]
```diff
@@ -300,10 +300,10 @@ VkInstance createInstance(ExtensionSet const& requiredExts) {
 }
 
 VkDevice createLogicalDevice(VkPhysicalDevice physicalDevice,
-        VkPhysicalDeviceFeatures const& features, uint32_t graphicsQueueFamilyIndex,
-        ExtensionSet const& deviceExtensions) {
+        VkPhysicalDeviceFeatures2 const& features, uint32_t graphicsQueueFamilyIndex,
+    uint32_t protectedGraphicsQueueFamilyIndex, ExtensionSet const& deviceExtensions) {
```
Comment (poweifeng) [inline filament/backend/src/vulkan/platform/VulkanPlatform.cpp:304]: spacing here.

File: filament/backend/include/backend/platforms/VulkanPlatform.h [hunk 3]
```diff
@@ -88,6 +88,7 @@ class VulkanPlatform : public Platform, utils::PrivateImplementation<VulkanPlatf
         VkFormat colorFormat = VK_FORMAT_UNDEFINED;
         VkFormat depthFormat = VK_FORMAT_UNDEFINED;
         VkExtent2D extent = {0, 0};
+        bool protection = false;
```
Comment (poweifeng) [inline filament/backend/include/backend/platforms/VulkanPlatform.h:91]: nit: prefer `protected`

File: filament/backend/src/vulkan/VulkanTexture.cpp [hunk 4]
```diff
@@ -232,6 +235,9 @@ VulkanTexture::VulkanTexture(VkDevice device, VkPhysicalDevice physicalDevice,
         imageInfo.arrayLayers = depth * 6;
         imageInfo.extent.depth = 1;
     }
+    if (state->mIsProtected) {
```
Comment (poweifeng) [inline filament/backend/src/vulkan/VulkanTexture.cpp:238]: nit: for consistency, write `any(usage & TextureUsage::PROTECTED)`

File: filament/backend/src/vulkan/VulkanTexture.h [hunk 5]
```diff
@@ -35,7 +35,7 @@ class VulkanResourceAllocator;
 struct VulkanTextureState : public VulkanResource {
     VulkanTextureState(VkDevice device, VmaAllocator allocator, VulkanCommands* commands,
             VulkanStagePool& stagePool, VkFormat format, VkImageViewType viewType, uint8_t levels,
-            uint8_t layerCount, VulkanLayout defaultLayout);
+            uint8_t layerCount, VulkanLayout defaultLayout, bool isProtected = false);
```
Comment (poweifeng) [inline filament/backend/src/vulkan/VulkanTexture.h:38]: prefer no default value here for `isProtected`

File: filament/backend/src/vulkan/platform/VulkanPlatform.cpp [hunk 6]
```diff
@@ -727,14 +773,23 @@ Driver* VulkanPlatform::createDriver(void* sharedContext,
 
     mImpl->mDevice
             = mImpl->mDevice == VK_NULL_HANDLE ? createLogicalDevice(mImpl->mPhysicalDevice,
-                      context.mPhysicalDeviceFeatures, mImpl->mGraphicsQueueFamilyIndex, deviceExts)
+                      context.mPhysicalDeviceFeatures, mImpl->mGraphicsQueueFamilyIndex, mImpl->mProtectedGraphicsQueueFamilyIndex, deviceExts)
```
Comment (poweifeng) [inline filament/backend/src/vulkan/platform/VulkanPlatform.cpp:776]: line too long, should <= 100 chars

File: filament/backend/src/vulkan/platform/VulkanPlatform.cpp [hunk 7]
```diff
@@ -698,10 +730,20 @@ Driver* VulkanPlatform::createDriver(void* sharedContext,
                       : mImpl->mGraphicsQueueFamilyIndex;
     assert_invariant(mImpl->mGraphicsQueueFamilyIndex != INVALID_VK_INDEX);
 
+    // We know we need to allocate the protected version of the VK objects
+    context.mProtectedMemorySupported = static_cast<bool>(queryProtectedMemoryFeatures.protectedMemory);
+    if (context.mProtectedMemorySupported) {
+        mImpl->mProtectedGraphicsQueueFamilyIndex
+            = mImpl->mProtectedGraphicsQueueFamilyIndex == INVALID_VK_INDEX
```
Comment (poweifeng) [inline filament/backend/src/vulkan/platform/VulkanPlatform.cpp:737]: wouldn't this always be true at this point?

File: filament/backend/src/vulkan/platform/VulkanPlatform.cpp [hunk 8]
```diff
@@ -314,22 +314,29 @@ VkDevice createLogicalDevice(VkPhysicalDevice physicalDevice,
...
+    // Protected queue
+    deviceQueueCreateInfo[1].sType = VK_STRUCTURE_TYPE_DEVICE_QUEUE_CREATE_INFO;
+    deviceQueueCreateInfo[1].flags = VK_DEVICE_QUEUE_CREATE_PROTECTED_BIT;
+    deviceQueueCreateInfo[1].queueFamilyIndex = protectedGraphicsQueueFamilyIndex;
+    deviceQueueCreateInfo[1].queueCount = 1;
+    deviceQueueCreateInfo[1].pQueuePriorities = &queuePriority[0];
+
     deviceCreateInfo.sType = VK_STRUCTURE_TYPE_DEVICE_CREATE_INFO;
-    deviceCreateInfo.queueCreateInfoCount = 1;
+    deviceCreateInfo.queueCreateInfoCount = protectedGraphicsQueueFamilyIndex == INVALID_VK_INDEX ? 1 : 2;
```
Comment (poweifeng) [inline filament/backend/src/vulkan/platform/VulkanPlatform.cpp:329]: how about we pull this into a boolean like `bool const hasProtectedQueue = protectedGraphicsQueueFamilyIndex == INVALID_VK_INDEX`?

File: filament/backend/src/vulkan/platform/VulkanPlatform.cpp [hunk 9]
```diff
@@ -360,6 +367,16 @@ VkDevice createLogicalDevice(VkPhysicalDevice physicalDevice,
         pNext = &multiview;
     }
 
+    VkPhysicalDeviceProtectedMemoryFeatures protectedMemory = {
+        .sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_PROTECTED_MEMORY_FEATURES,
+    };
+    if (protectedGraphicsQueueFamilyIndex != INVALID_VK_INDEX) {
```
Comment (poweifeng) [inline filament/backend/src/vulkan/platform/VulkanPlatform.cpp:373]: this can be `if (hasProtectedQueue)`

**Comment by poweifeng** [comment]
> As discussed this is the first phase of content protection.

Can you give more description on what this PR is doing?

File: filament/backend/src/vulkan/platform/VulkanPlatform.cpp (hunk 7 above)
Comment (phoenixxxx) [reply filament/backend/src/vulkan/platform/VulkanPlatform.cpp:737]: Not necessarily on all platforms. This is queried from the driver.

**Comment by phoenixxxx** [comment]
This PR is the first part of a two-part chunk of work for supporting protected content rendering in the Vulkan path. 
- Create the protection flag (and accessors) in the platform
- Create the protection flag (and accessors) in the context
- Create the protection flag (and accessors) in the texture
- Create the protection flag (and accessors) in the swap chain

We also pass the protection in the swap chain bundle when creating a render target from individual vkImages. Finally we create a protected render queue should the GPU support it.

File: filament/backend/src/vulkan/platform/VulkanPlatform.cpp [hunk 10]
```diff
@@ -406,13 +424,13 @@ FixedCapacityVector<VkQueueFamilyProperties> getPhysicalDeviceQueueFamilyPropert
     return queueFamiliesProperties;
 }
 
-uint32_t identifyGraphicsQueueFamilyIndex(VkPhysicalDevice physicalDevice) {
+uint32_t identifyGraphicsQueueFamilyIndex(VkPhysicalDevice physicalDevice, VkQueueFlags flags = VK_QUEUE_GRAPHICS_BIT) {
```
Comment (poweifeng) [inline filament/backend/src/vulkan/platform/VulkanPlatform.cpp:427]: no default for `flags` please

File: filament/backend/include/backend/platforms/VulkanPlatform.h (hunk 3 above)
Comment (phoenixxxx) [reply filament/backend/include/backend/platforms/VulkanPlatform.h:91]: that's a keyword.

Comment (poweifeng) [reply filament/backend/include/backend/platforms/VulkanPlatform.h:91]: ah yes.  How about `isProtected`

File: filament/backend/src/vulkan/platform/VulkanPlatform.cpp [hunk 11]
```diff
@@ -473,7 +491,7 @@ VkPhysicalDevice selectPhysicalDevice(VkInstance instance,
         // Does the device have any command queues that support graphics?
         // In theory, we should also ensure that the device supports presentation of our
         // particular VkSurface, but we don't have a VkSurface yet, so we'll skip this requirement.
-        if (identifyGraphicsQueueFamilyIndex(candidateDevice) == INVALID_VK_INDEX) {
+        if (identifyGraphicsQueueFamilyIndex(candidateDevice, VK_QUEUE_GRAPHICS_BIT) == INVALID_VK_INDEX) {
```
Comment (poweifeng) [inline filament/backend/src/vulkan/platform/VulkanPlatform.cpp:494]: nit: line too long

File: filament/backend/src/vulkan/platform/VulkanPlatform.cpp [hunk 12]
```diff
@@ -686,22 +707,43 @@ Driver* VulkanPlatform::createDriver(void* sharedContext,
...
-    vkGetPhysicalDeviceProperties(mImpl->mPhysicalDevice, &context.mPhysicalDeviceProperties);
-    vkGetPhysicalDeviceFeatures(mImpl->mPhysicalDevice, &context.mPhysicalDeviceFeatures);
+    vkGetPhysicalDeviceProperties2(mImpl->mPhysicalDevice, &context.mPhysicalDeviceProperties);
+    vkGetPhysicalDeviceFeatures2(mImpl->mPhysicalDevice, &context.mPhysicalDeviceFeatures);
     vkGetPhysicalDeviceMemoryProperties(mImpl->mPhysicalDevice, &context.mMemoryProperties);
 
     mImpl->mGraphicsQueueFamilyIndex
             = mImpl->mGraphicsQueueFamilyIndex == INVALID_VK_INDEX
-                      ? identifyGraphicsQueueFamilyIndex(mImpl->mPhysicalDevice)
+                      ? identifyGraphicsQueueFamilyIndex(mImpl->mPhysicalDevice, VK_QUEUE_GRAPHICS_BIT)
```
Comment (poweifeng) [inline filament/backend/src/vulkan/platform/VulkanPlatform.cpp:730]: over 100 chars

File: filament/backend/src/vulkan/platform/VulkanPlatform.cpp [hunk 13]
```diff
@@ -686,22 +707,43 @@ Driver* VulkanPlatform::createDriver(void* sharedContext,
...
                       : mImpl->mGraphicsQueueFamilyIndex;
     assert_invariant(mImpl->mGraphicsQueueFamilyIndex != INVALID_VK_INDEX);
 
+    // We know we need to allocate the protected version of the VK objects
+    context.mProtectedMemorySupported = static_cast<bool>(queryProtectedMemoryFeatures.protectedMemory);
```
Comment (poweifeng) [inline filament/backend/src/vulkan/platform/VulkanPlatform.cpp:735]: over 100 chars

File: filament/backend/src/vulkan/platform/VulkanPlatform.cpp [hunk 14]
```diff
@@ -711,6 +753,10 @@ Driver* VulkanPlatform::createDriver(void* sharedContext,
     mImpl->mGraphicsQueueIndex
             = mImpl->mGraphicsQueueIndex == INVALID_VK_INDEX ? 0 : mImpl->mGraphicsQueueIndex;
 
+    // Applying the same logic to the protected queue index (Not sure about shared context and protection)
+    mImpl->mProtectedGraphicsQueueIndex
+        = mImpl->mProtectedGraphicsQueueIndex == INVALID_VK_INDEX ? 0 : mImpl->mProtectedGraphicsQueueIndex;
```
Comment (poweifeng) [inline filament/backend/src/vulkan/platform/VulkanPlatform.cpp:758]: over 100 chars

**Pull request merged**